src/server/pps/server/master.go

```diff
@@ -158,6 +166,31 @@ eventLoop:
 	}
 }
 
+func (m *ppsMaster) attemptStep(ctx context.Context, e *pipelineEvent) error {
+	var errCount int
+	return backoff.RetryNotify(func() error {
+		// Create/Modify/Delete pipeline resources as needed per new state
+		return m.step(e.pipeline, e.etcdVer, e.etcdRev)
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
```

Comment:
I don't have a sense of what our real-world error distribution looks like here - clearly after my change we're a little less tolerant to errors since there are fewer overall retry attempts, but maybe that doesn't matter. Limiting the total time spent retrying sounds good. The main difference that would cause is by default ExponentialBackoff skips calling notify if it hits a timeout, which means master wouldn't attempt to fail the pipeline, but maybe working around that by moving the fail step outside of the backoff isn't so bad. Perhaps I'm worrying too much, though, since as you point out the poller will eventually fix things.